### PR TITLE
Fix for #189 variable substitution OutOfMemory error

### PR DESCRIPTION
**Pull request by omnibs**
The regex to prevent throwing errors at unset variables inside `VariableSubstitutionPreprocessor` is too complex and susceptible to stack overflow and out of memory errors.

I've rewritten what it does without regex, parsing the comments and marking the index ranges, with tests and all.

One thing I've noticed writing this is that this method of dealing with comments is very error prone if scripts contain strings with `/*` or `--`. 

It's an issue in the current code and also in this new one I'm proposing. In order to handle that we'd have to parse strings in the query also. I'm willing to try that out in a second PR, since this is already an issue in the master branch.

### CODE REVIEW

**Comment by dazinator** [comment]
Not positive if this is related but could the `SqlCommandReader` class that we have help here at all? It's basically a reader that is intelligent enough to read SQL files, with semantics for reading comment blocks (--, */) and splitting on GO statements that are not within comment blocks. We use it so that we can read 1 "command" at a time, even if the script contains multiple commands with dash dash or slash star comments. We used to use a regex for splitting on GO but it was pretty ugly imho and not very robust!

**Comment by omnibs** [comment]
Good idea @dazinator, seems like the perfect candidate. 

I've seen we have one for SqlServer that apparently is also used for SqlCe, one for MySql. Do you have any idea what kind of trouble could we run into using that class for pgsql or firebird?

**Comment by dazinator** [comment]
I had a quick review of the reader. Because both those systems use the sql ansi standard for comments format (single line and multiling) and for string literals, the bulk of the logic in the reader should work fine with postgres and firebird.

However one minor problem I did spot is that the reader currently assumes that [] are used to escape object names. The reader reads past bracketed values so it doesn't falsely detect things like "GO" within object names. This works fine for the MSSQL etc but Firebird uses double quotes, ". not [].

So basically, one tweak we'd probably have to make is to allow those characters to be configurable / overriden rather than hardcoded to [] and then that should be ok.

**Comment by omnibs** [comment]
Work work work... finally had some time to get back to this issue.

I've taken a look at `SqlCommandReader`. In order to use it I'd have to change it in ways that make no sense for a `SqlCommandReader`. 

I'm trying my hand at breaking it apart. Extracting a `SqlParser`, implementing the original `SqlCommandReader` over it and porting variable substitution to that.

**Comment by JakeGinnivan** [comment]
Hey guys, I am going to push up the netcore branch tonight. Do you recon you will have time over the next few days to look a the SqlCommandReader approach and get into the v4 branch?

**Comment by omnibs** [comment]
Jake, thanks for the reminder, just made some time to come back and finish this.

I had a hard time figuring out why we needed to check [this](https://github.com/DbUp/DbUp/blob/master/src/DbUp/Support/SqlServer/SqlCommandReader.cs#L431-L434) on `ReadDelimiter`:

```C#
if (Read(buffer, 0, Delimiter.Length - 1) != Delimiter.Length - 1)
{
    return true;
}
```

When we check for delimiters we already peek ahead for the delimiter's length [here](https://github.com/DbUp/DbUp/blob/master/src/DbUp/Support/SqlServer/SqlCommandReader.cs#L336) on `IsBeginningOfDelimiter`, so I believe that `if` could never be true.

```C#
TryPeek(Delimiter.Length - 1, out result)
```

Could you guys double check I haven't missed anything?

Jake, I can squash these commits if you give me a thumbs up, I've just left them here for clarity.

File: src/DbUp/Engine/Preprocessors/VariableSubstitutionPreprocessor.cs [hunk 1]
```diff
@@ -29,47 +33,82 @@ public VariableSubstitutionPreprocessor(IDictionary<string, string> variables)
         /// <param name="contents"></param>
         public string Process(string contents)
         {
-            Regex regexObj = new Regex(@"/\*(?>(?:(?!\*/|/\*).)*)(?>(?:/\*(?>(?:(?!\*/|/\*).)*)\*/(?>(?:(?!\*/|/\*).)*))*).*?\*/|--.*?\r?[\n]", RegexOptions.Singleline);
-            Match commentMatch = regexObj.Match(contents);
-
-            return tokenRegex.Replace(contents, match => ReplaceToken(match, variables, commentMatch));
+            using (var parser = new VariableSubstitutionSqlParser(contents))
```
Comment (dazinator) [inline src/DbUp/Engine/Preprocessors/VariableSubstitutionPreprocessor.cs]: Can this line now be removed? https://github.com/DbUp/DbUp/pull/190/files#diff-d9ae76871876bd2c951280aa2608d84aR19

Comment (omnibs) [inline src/DbUp/Engine/Preprocessors/VariableSubstitutionPreprocessor.cs]: Totally, slipped past me!

File: src/DbUp/Support/SqlServer/SqlParser.cs [hunk 2]
```diff
@@ -0,0 +1,518 @@
...
+            var previousChar = CurrentChar;
+            var buffer = new char[Delimiter.Length - 1];
+
+            // read the rest of the delimiter
+            Read(buffer, 0, Delimiter.Length - 1);
+
+            // Support terminator.
+            var peekChar = PeekChar();
+            char? terminator = null;
+            if (peekChar == ';' || peekChar == '\0')
```
Comment (dazinator) [inline src/DbUp/Support/SqlServer/SqlParser.cs]: consider changing `\0` to `NullChar` constant instead?

Comment (omnibs) [inline src/DbUp/Support/SqlServer/SqlParser.cs]: Done! =]

**Comment by dazinator** [comment]
I think these are all good changes! Nice one :)

**Comment by omnibs** [comment]
Thanks for the review!

Just pushed the suggestions =]

File: src/DbUp/Engine/Preprocessors/VariableSubstitutionPreprocessor.cs [hunk 3]
```diff
@@ -29,47 +33,82 @@ public VariableSubstitutionPreprocessor(IDictionary<string, string> variables)
...
+            {
+                return parser.ReplaceVariables(variables);
+            }
         }
 
-        private static bool IsInComment(Match commentMatch, Match variableMatch)
+        private class VariableSubstitutionSqlParser : SqlParser
```
Comment (dazinator) [inline src/DbUp/Engine/Preprocessors/VariableSubstitutionPreprocessor.cs]: Open question: Do you think there would be any value in making this public? I was just thinking would people potentially want to derive from this class to customer variable substitution behavior with their own preprocessors?

Comment (omnibs) [inline src/DbUp/Engine/Preprocessors/VariableSubstitutionPreprocessor.cs]: Good question. Since the previous VariableSubstitutionPreprocessor wasn't very customizable I didn't think much about this.

I'll expose it with the `'$'` assigned to a virtual property, so a simple use case would be overriding the variable delimiter.

I won't expose the option to change the delimiter on the `VariableSubstitutionPreprocessor` so we don't have to worry about giving any guarantees as to what delimiters can be used without breaking `SqlParser`.

**Comment by dazinator** [comment]
Nice one.

> When we check for delimiters we already peek ahead for the delimiter's length here on IsBeginningOfDelimiter, so I believe that if could never be true.

I think you are correct, good spot sir! I'd be inclined to remove the redundant check. There is plenty of test coverage for the `SqlCommandParser` so we should soon pick up on any regressions caused by such a change :)

**Comment by omnibs** [comment]
Exposed the `VariableSubstitutionSqlParser`. 

The `/Preprocessors` directory might not be the best place for it but IDK, `Support/SqlServer` seemed too far away from `VariableSubstitutionPreprocessor`.

As always, forgot to run the public API tests. Oh, the value of CI, right? =p

**Comment by dazinator** [comment]
All LGTM :+1: great work @omnibs 
Think we are ready for @JakeGinnivan to cast his magic.

**Comment by JakeGinnivan** [comment]
Thanks guys. I am working on a proper automated deploy for Shouldly. Once I get 2.8.0 out the door using that I will switch to getting v4 beta out using a similar technique. Then we can all push the button to release :)

**Comment by JakeGinnivan** [comment]
Likely I will get to this early next week after I finish off shouldly then the DbUp netcore branch :)
